src/seccomp_fd_notify.c

```diff
@@ -0,0 +1,164 @@
+#include <sys/prctl.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <linux/audit.h>
+#include <linux/seccomp.h>
+#include <linux/filter.h>
+#include <sys/syscall.h>
+#include <stddef.h>
+
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <attr/xattr.h>
+
+#include "seccomp_fd_notify.h"
+
+#define PRINT_WARNING(...)  { fprintf(stderr, __VA_ARGS__); fprintf(stderr, "\n"); }
+#define PRINT_INFO(...)  { fprintf(stderr, __VA_ARGS__); fprintf(stderr, "\n"); }
+
+
+int sendfd(int sockfd, int fd) {
+	struct msghdr msgh;
+	struct iovec iov;
+	int data;
+	struct cmsghdr *cmsgp;
+	union {
+		char   buf[CMSG_SPACE(sizeof(int))];
+		/* Space large enough to hold an 'int' */
+		struct cmsghdr align;
+	} controlMsg;
+
+	msgh.msg_name = NULL;
+	msgh.msg_namelen = 0;
+	msgh.msg_iov = &iov;
+	msgh.msg_iovlen = 1;
+	iov.iov_base = &data;
+	iov.iov_len = sizeof(int);
+	data = 12345;
+
+	msgh.msg_control = controlMsg.buf;
+	msgh.msg_controllen = sizeof(controlMsg.buf);
+
+	cmsgp = CMSG_FIRSTHDR(&msgh);
+	cmsgp->cmsg_level = SOL_SOCKET;
+	cmsgp->cmsg_type = SCM_RIGHTS;
+	cmsgp->cmsg_len = CMSG_LEN(sizeof(int));
+	memcpy(CMSG_DATA(cmsgp), &fd, sizeof(int));
+
+	if (sendmsg(sockfd, &msgh, 0) == -1)
+		return -1;
+
+	return 0;
+}
+
+static int
+seccomp(unsigned int operation, unsigned int flags, void *args)
+{
+	return syscall(__NR_seccomp, operation, flags, args);
+}
+
+
+/* The following is the x86-64-specific BPF boilerplate code for checking that
+   the BPF program is running on the right architecture + ABI. At completion
+   of these instructions, the accumulator contains the system call number. */
+
+/* For the x32 ABI, all system call numbers have bit 30 set */
+
+#define X32_SYSCALL_BIT         0x40000000
+
+#define X86_64_CHECK_ARCH_AND_LOAD_SYSCALL_NR \
+	BPF_STMT(BPF_LD | BPF_W | BPF_ABS, \
+		(offsetof(struct seccomp_data, arch))), \
+	BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K, AUDIT_ARCH_X86_64, 0, 2), \
+	BPF_STMT(BPF_LD | BPF_W | BPF_ABS, \
+		 (offsetof(struct seccomp_data, nr))), \
+	BPF_JUMP(BPF_JMP | BPF_JGE | BPF_K, X32_SYSCALL_BIT, 0, 1), \
+	BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_KILL_PROCESS)
+
+/* installNotifyFilter() installs a seccomp filter that generates user-space
+   notifications (SECCOMP_RET_USER_NOTIF) when the process calls mkdir(2); the
+   filter allows all other system calls.
+
+   The function return value is a file descriptor from which the user-space
+   notifications can be fetched. */
+
+static int installNotifyFilter(void) {
+	struct sock_filter filter[] = {
+		X86_64_CHECK_ARCH_AND_LOAD_SYSCALL_NR,
```

Comment:
the define up above is confusing. Why not just include it

Reply:
Heh, I guess it is always using the X86_64 thing anyway. Sure.

Reply:
we need to make sure we support x32 and x86-64.